### PR TITLE
Update graphql, events & extending docs for Laravel and v4

### PR DESCRIPTION
**Pull request by vinceAmstoutz**
The initial aim was to update the `core/graphql.md`, but  I soon realized, with @soyuka's help, that some parts didn't work like that anymore in API Platform 4 (for example the "stages").

So we decided to move and rewrite some parts in `core/extending.md`  and add a warning in `core/events.md` documentations.

### CODE REVIEW

File: core/extending.md [hunk 1]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
 To replace existing API Platform services with your decorators, [check out how to decorate services](https://symfony.com/doc/current/service_container/service_decoration.html).
 
 <p align="center" class="symfonycasts"><a href="https://symfonycasts.com/screencast/api-platform-security/service-decoration?cid=apip"><img src="../symfony/images/symfonycasts-player.png" alt="Service Decoration screencast"><br>Watch the Service Decoration screencast</a></p>
+
+## System providers and processors
```
Comment (dunglas) [inline core/extending.md:40]:
```suggestion
## System Providers and Processors
```

File: core/extending.md [hunk 2]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+
+### Symfony Access Checker Provider
+
+When using Symfony, the access checker provider is used at three different stages:
+- `api_platform.state_provider.access_checker.post_validate` decorates the ValidateProvider
+- `api_platform.state_provider.access_checker.post_deserialize` decorates the DeserializeProvider
+- `api_platform.state_provider.access_checker` decorates the ReadProvider
+
+
+### Decoration example
```
Comment (dunglas) [inline core/extending.md:67]:
```suggestion
### Decoration Example
```

File: core/extending.md [hunk 3]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+    C3 --> C4(ParameterProvider)
+    C4 --> C5(ValidateProcessor)
+    C5 --> C6(WriteProcessor)
+    C6 --> C7(SerializeProcessor)
+```
+
+### Symfony Access Checker Provider
+
+When using Symfony, the access checker provider is used at three different stages:
+- `api_platform.state_provider.access_checker.post_validate` decorates the ValidateProvider
```
Comment (dunglas) [inline core/extending.md:62]: Maybe add backticks around class names?

File: core/extending.md [hunk 4]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+
+```php
+<?php
+namespace App\State;
+
+use ApiPlatform\State\ProcessorInterface;
+
+final class CustomRespondProcessor implements ProcessorInterface
+{
+    public function __construct(private readonly ProcessorInterface $processor){}
```
Comment (dunglas) [inline core/extending.md:81]: CS to fix

File: core/graphql.md [hunk 5]
```diff
@@ -10,23 +10,23 @@ Once enabled, you have nothing to do: your schema describing your API is automat
...
 
-*Note:* If you used [Symfony Flex to install API Platform](../symfony/index.md#using-symfony-flex-and-composer-advanced-users), URLs will be prefixed with `/api` by default. For example, the GraphQL endpoint will be: `https://localhost:8443/api/graphql`.
+> [!NOTE]
+> If you used [the Symfony Variant thanks to Symfony Flex](../symfony/index.md#using-symfony-flex-and-composer-advanced-users) or the Laravel variant, URLs will be prefixed with `/api` by default. For example, the GraphQL endpoint will be: `https://localhost:8443/api/graphql`.
 
 ## Changing Location of the GraphQL Endpoint
-
 Sometimes you may want to have the GraphQL endpoint at a different location. This can be done by manually configuring the GraphQL controller.
 
+### Symfony config for routes
```
Comment (dunglas) [inline core/graphql.md:27]:
```suggestion
### Symfony Config for Routes
```

Comment (dunglas) [inline core/graphql.md:27]:
```suggestion
### Symfony Routes
```

File: core/graphql.md [hunk 6]
```diff
@@ -10,23 +10,23 @@ Once enabled, you have nothing to do: your schema describing your API is automat
...
+
+Using the Symfony variant we can do this modification by adding the following code:
```
Comment (dunglas) [inline core/graphql.md:29]:
```suggestion
Using the Symfony variant we can do this modification by adding the following code:

```

File: core/graphql.md [hunk 7]
```diff
@@ -37,14 +37,29 @@ api_graphql_entrypoint:
 
 Change `/api/graphql` to the URI you wish the GraphQL endpoint to be accessible on.
 
+### Laravel config for routes
```
Comment (dunglas) [inline core/graphql.md:40]:
```suggestion
### Laravel Routes
```

File: core/graphql.md [hunk 8]
```diff
@@ -37,14 +37,29 @@ api_graphql_entrypoint:
...
 ## GraphiQL
 
-If Twig is installed in your project, go to the GraphQL endpoint with your browser. You will see a nice interface provided by GraphiQL to interact with your API.
+Go to the GraphQL endpoint with your browser, you will see a nice interface provided by GraphiQL to interact with your API.
 
 The GraphiQL IDE can also be found at `/graphql/graphiql`.
 
 If you need to disable it, it can be done in the configuration:
 
+### Disable GraphiQL with Symfony
```
Comment (dunglas) [inline core/graphql.md:62]:
```suggestion
### Disabling GraphiQL with Symfony
```

File: core/graphql.md [hunk 9]
```diff
@@ -54,9 +69,26 @@ api_platform:
 # ...
 ```
 
+### Disable GraphiQL with Laravel
```
Comment (dunglas) [inline core/graphql.md:72]:
```suggestion
### Disabling GraphiQL with Laravel

```

**Review by dunglas — COMMENTED** [review]
This looks great!

File: core/extending.md [hunk 10]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+    C2 --> C3(DeserializeProvider)
+    C3 --> C4(ParameterProvider)
+    C4 --> C5(ValidateProcessor)
+    C5 --> C6(WriteProcessor)
+    C6 --> C7(SerializeProcessor)
+```
+
+### Symfony Access Checker Provider
+
+When using Symfony, the access checker provider is used at three different stages:
```
Comment (soyuka) [inline core/extending.md:61]: This is only for rest I think

File: core/extending.md [hunk 11]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+<?php
+namespace App\State;
+
+use ApiPlatform\State\ProcessorInterface;
+
+final class CustomRespondProcessor implements ProcessorInterface
+{
+    public function __construct(private readonly ProcessorInterface $processor) {}
+
+    public function __invoke($data, string $resourceClass, string $operationName, array $context)
```
Comment (soyuka) [inline core/extending.md:83]: Can you implement `process` instead?

File: core/extending.md [hunk 12]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+use Illuminate\Contracts\Foundation\Application;
+use Illuminate\Support\ServiceProvider;
+
+class AppServiceProvider extends ServiceProvider
+{
+    public function register(): void
+    {
+        $this->app->singleton(CustomRespondProcessor::class, function (Application $app) {
+            return new CustomRespondProcessor();
+        });
```
Comment (soyuka) [inline core/extending.md:144]: The extend below should be sufficient

File: core/extending.md [hunk 13]
```diff
@@ -36,3 +36,120 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+        
+        $this->app->extend(RespondProcessor::class, function (RespondProcessor $respondProcessor, Application $app) {
```
Comment (soyuka) [inline core/extending.md:146]: Remove $app

**Review by soyuka — COMMENTED** [review]
Nice, on graphql we also need to document how to secure routes using a middleware and probably also document policies.

File: core/extending.md (hunk 10 above)
Comment (vinceAmstoutz) [reply core/extending.md:61]: For this one, do you want to add somewhere that it's only for REST?

**Comment by vinceAmstoutz** [comment]
> Nice, on graphql we also need to document how to secure routes using a middleware and probably also document policies.

Thanks, please let me know more about what's missing before we can merge.

File: core/extending.md (hunk 10 above)
Comment (soyuka) [reply core/extending.md:61]: just add a note that this is not available in graphql?

File: core/events.md [hunk 14]
```diff
@@ -1,9 +1,12 @@
...
-In API Platform 3.2 you may need `event_listeners_backward_compatibility_layer: true` to keep event listeners activated.
+> [!WARNING]
+> In API Platform 4.0 with Symfony, you need `use_symfony_listeners: true` to activate event listeners.
 
-Note: using Kernel event with API Platform should be mostly limited to tweaking the generated HTTP response. Also, GraphQL is **not supported**.
-[For most use cases, better extension points, working both with REST and GraphQL, are available](extending.md).
+---
+
+> [!NOTE]
+> Using Kernel event with API Platform should be mostly limited to tweaking the generated HTTP response. Also, GraphQL is **not supported**.
```
Comment (soyuka) [inline core/events.md:9]: Can you add:

We recommend to use [System providers and processors](./extending.md) to extend API Platform internals.

File: core/extending.md [hunk 15]
```diff
@@ -36,3 +36,116 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+```
+
+### Symfony Access Checker Provider
+
+When using Symfony, the access checker provider is used at three different stages:
+- `api_platform.state_provider.access_checker.post_validate` decorates the `ValidateProvider`
+- `api_platform.state_provider.access_checker.post_deserialize` decorates the `DeserializeProvider`
+- `api_platform.state_provider.access_checker` decorates the `ReadProvider`
+
+
```
Comment (soyuka) [inline core/extending.md:66]: > [!NOTE]
> For graphql use: `api_platform.graphql.state_provider.access_checker.post_deserialize`, `api_platform.graphql.state_provider.access_checker.post_validate`, `api_platform.graphql.state_provider.validate` and
 `api_platform.graphql.state_provider.access_checker.after_resolver`

File: core/events.md (hunk 14 above)
Comment (soyuka) [inline core/events.md:9]:
```suggestion
> Using Kernel event with API Platform should be mostly limited to tweaking the generated HTTP response. Also, GraphQL is **not supported**. We recommend to use [System providers and processors](https://github.com/api-platform/docs/pull/2026/extending.md) to extend API Platform internals.
```

File: core/extending.md [hunk 16]
```diff
@@ -36,3 +36,116 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+        $this->processor->process($data, $operation, $uriVariables, $context);
+
+        // You can add post-write code here.
+
+        return $writtenObject;
+    }
+}
+```
+
+Now you should decorate the RespondProcessor with the CustomRespondProcessor using Symfony or Laravel:
```
Comment (soyuka) [inline core/extending.md:97]:
```suggestion
Now decorate the `RespondProcessor` with the `CustomRespondProcessor` using Symfony or Laravel:
```

File: core/extending.md [hunk 17]
```diff
@@ -36,3 +36,116 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+
+### Decorating the RespondProcessor with Symfony
```
Comment (soyuka) [inline core/extending.md:99]:
```suggestion
### Symfony Processor Decoration
```

File: core/extending.md [hunk 18]
```diff
@@ -36,3 +36,116 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+# api/config/services.yaml
+services:
+    # ...
+    App\State\CustomRespondProcessor:
+        decorates: api_platform.state.processor.respond_processor
+```
+
+And that's it!
+
+### Decorating the RespondProcessor with Laravel
```
Comment (soyuka) [inline core/extending.md:127]:
```suggestion
### Laravel Processor Decoration
```

File: core/extending.md [hunk 19]
```diff
@@ -36,3 +36,116 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+    public function register(): void
+    {
+        $this->app->extend(RespondProcessor::class, function (RespondProcessor $respondProcessor) {
+            return new CustomRespondProcessor($respondProcessor);
+        });
+    }
+
+    public function boot(): void
+    {
+    }
```
Comment (soyuka) [inline core/extending.md:149]:
```suggestion
```

File: core/graphql.md [hunk 20]
```diff
@@ -10,23 +10,25 @@ Once enabled, you have nothing to do: your schema describing your API is automat
 
 ## Enabling GraphQL
 
-To enable GraphQL and its IDE (GraphiQL and GraphQL Playground) in your API, simply require the [graphql-php](https://webonyx.github.io/graphql-php/) package using Composer and clear the cache one more time:
+To enable GraphQL and its IDE (GraphiQL and GraphQL Playground) in your API, simply require the [API Platform GraphQL](https://github.com/api-platform/graphql) package using Composer:
```
Comment (soyuka) [inline core/graphql.md:13]:
```suggestion
To enable GraphQL and its IDE (GraphiQL and GraphQL Playground) in your API, simply require the `api-platform/graphql` package using Composer:
```

File: core/graphql.md [hunk 21]
```diff
@@ -54,9 +71,27 @@ api_platform:
 # ...
 ```
 
+### Disabling GraphiQL with Laravel
+
+```php
+<?php
+//api/config/api-platform.php
```
Comment (soyuka) [inline core/graphql.md:78]:
```suggestion
// app/config/api-platform.php
```

I think this is not inside `api`, can you check?

File: core/graphql.md [hunk 22]
```diff
@@ -82,9 +131,27 @@ api_platform:
...
+### Disable GraphQL Playground with Laravel
+
+```php
+<?php
+//api/config/api-platform.php
+return [
+    // ....
+    'graphql' => [
+        'graphql_playground' => [
+            'enabled' => false,
```
Comment (soyuka) [inline core/graphql.md:143]: this is not yet supported inside Laravel we should add it (basically we only need to register the controller)

File: core/graphql.md [hunk 23]
```diff
@@ -861,72 +881,24 @@ API Platform resolves the queries and mutations by using its own **resolvers**.
...
-    {
-        // You can add pre-write code here.
+See the [Extending API Platform](extending.md) documentation for more information.
 
-        // Call the decorated write stage (this syntax calls the __invoke method).
-        $writtenObject = ($this->writeStage)($data, $resourceClass, $operationName, $context);
+### Disabling system providers and processors
 
-        // You can add post-write code here.
+If you need to, you can disable some states providers and state processors done by the resolvers, for instance if you don't want your data to be validated.
```
Comment (soyuka) [inline core/graphql.md:888]:
```suggestion
If you need to, you can disable some states providers and state processors, for instance if you don't want your data to be validated.
```

File: core/graphql.md [hunk 24]
```diff
@@ -2243,6 +2261,37 @@ return function(ContainerConfigurator $configurator) {
...
+use Illuminate\Support\ServiceProvider;
+
+class AppServiceProvider extends ServiceProvider
+{
+    public function register(): void
+    {
+        $this->app->singleton(ErrorHandlerDecorated::class, function (Application $app) {
+            return new ErrorHandlerDecorated();
+        });
+
```
Comment (soyuka) [inline core/graphql.md:2283]:
```suggestion
```

File: core/graphql.md [hunk 25]
```diff
@@ -2243,6 +2261,37 @@ return function(ContainerConfigurator $configurator) {
...
+        });
+        
+        $this->app->extend(ErrorHandler::class, function (ErrorHandler $errorHandler, Application $app) {
+            return new ErrorHandlerDecorated($errorHandler);
+        });
+    }
+
+    public function boot(): void
+    {
+    }
```
Comment (soyuka) [inline core/graphql.md:2291]:
```suggestion
```

File: core/graphql.md [hunk 26]
```diff
@@ -2538,8 +2621,42 @@ Your custom type is now registered and is available in the `TypesContainer`.
...
+class AppServiceProvider extends ServiceProvider
+{
+    public function register(): void
+    {
+        $this->app->tag([DateTimeType::class], TypeInterface::class);
+    }
+
+    public function boot(): void
+    {
+    }
```
Comment (soyuka) [inline core/graphql.md:2647]:
```suggestion
```

Really not sure but I think it's more readable no?

File: core/graphql.md [hunk 27]
```diff
@@ -2538,8 +2621,42 @@ Your custom type is now registered and is available in the `TypesContainer`.
...
+
+Your custom type is now registered and is available in the `TypesContainer`.
+
+To use it please [modify the extracted types](#modify-the-extracted-types) or use it directly in [custom queries](#custom-queries) or [custom mutations](#custom-mutations).
+
 ## Modify the Extracted Types
 
+> [!WARNING]
+> This part is only supported for API Platform with Symfony
+
```
Comment (soyuka) [inline core/graphql.md:2659]: we can extend that using laravel as well ?

File: core/graphql.md [hunk 28]
```diff
@@ -2606,7 +2723,10 @@ if (Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType()
 
 All `DateTimeInterface` properties will have the `DateTime` type in this example.
 
-## Changing the Serialization Context Dynamically
+## Changing the Serialization Context Dynamically (only for Symfony)
+
+> [!WARNING]
+> This part is only supported for API Platform with Symfony
```
Comment (soyuka) [inline core/graphql.md:2729]: this also works for laravel

File: core/graphql.md [hunk 29]
```diff
@@ -2651,19 +2771,20 @@ final class BookContextBuilder implements SerializerContextBuilderInterface
 
 ## Export the Schema in SDL
 
+> [!WARNING]
+> These commands are only available for API Platform with Symfony!
```
Comment (soyuka) [inline core/graphql.md:2775]:
```suggestion
> This command is not yet available with Laravel, you're welcome to contribute [on Github](github.com/api-platform/core)
```

**Review by soyuka — CHANGES_REQUESTED** [review]

File: core/extending.md [hunk 30]
```diff
@@ -36,3 +36,116 @@ For instance, if you want to send a mail after a resource has been persisted, bu
...
+### Decorating the RespondProcessor with Symfony
+
+With Symfony you can simply do that by adding the `#[AsDecorator]` attribute as following:
+
+```php
+namespace App\State;
+
+use ApiPlatform\State\ProcessorInterface;
+
+#[AsDecorator(decorates: 'api_platform.state.processor.respond_processor',)]
```
Comment (alanpoulain) [inline core/extending.md:108]:
```suggestion
#[AsDecorator(decorates: 'api_platform.state.processor.respond_processor')]
```

File: core/graphql.md [hunk 31]
```diff
@@ -266,9 +412,11 @@ final class BookCollectionResolver implements QueryCollectionResolverInterface
 }
 ```
 
+### Custom Queries config for Symfony
+
 If you use autoconfiguration (the default Symfony configuration) in your application, then you are done!
 
-Else, you need to tag your resolver like this:
+Else, you need to tag your resolver like this if you using Symfony without autoconfiguration :
```
Comment (alanpoulain) [inline core/graphql.md:419]:
```suggestion
Else, you need to tag your resolver like this if you are using Symfony without autoconfiguration :
```

File: core/graphql.md [hunk 32]
```diff
@@ -848,7 +867,8 @@ For instance, you could receive a JSON payload like this:
 
 ### Subscriptions Cache
 
-Internally, API Platform stores the subscriptions in a cache, using the [Symfony Cache](https://symfony.com/doc/current/cache.html).
+Internally, API Platform stores the subscriptions in a cache, using the [Symfony Cache](https://symfony.com/doc/current/cache.html) if you're using the Symfony variant
+or ÆPI Platform uses [Laravel cache](https://laravel.com/docs/cache) if you're use the Laravel variant.
```
Comment (alanpoulain) [inline core/graphql.md:871]:
```suggestion
or API Platform uses [Laravel cache](https://laravel.com/docs/cache) if you're using the Laravel variant.
```

File: core/graphql.md [hunk 33]
```diff
@@ -2263,6 +2312,9 @@ It's the case for a `HttpException` for which the `status` entry will be added u
 
 #### Custom Exception Normalizer
 
+> [!CAUTION]
+> Works only for API Platform with Symfony,
```
Comment (alanpoulain) [inline core/graphql.md:2312]:
```suggestion
> Works only for API Platform with Symfony
```

**Comment by vinceAmstoutz** [comment]
@soyuka @alanpoulain Reviewed changes resolved!

**Review by alanpoulain — APPROVED** [review]
Don't forget to squash!

File: core/graphql.md [hunk 34]
```diff
@@ -2263,6 +2293,9 @@ It's the case for a `HttpException` for which the `status` entry will be added u
 
 #### Custom Exception Normalizer
 
+> [!CAUTION]
+> Works only for API Platform with Symfony
+
```
Comment (soyuka) [inline core/graphql.md:2298]:
```suggestion
```

**Comment by soyuka** [comment]
Nice thanks!

**Pull request merged**